src/java.base/share/classes/java/util/DualPivotQuicksort.java

```diff
@@ -618,6 +636,126 @@ private static void pushDown(int[] a, int p, int value, int low, int high) {
         a[p] = value;
     }
 
+    /**
+     * Sorts the specified range of the array using radix sort.
+     *
+     * @param a the array to be sorted
+     * @param low the index of the first element, inclusive, to be sorted
+     * @param high the index of the last element, exclusive, to be sorted
+     * @return true if finally sorted, false otherwise
+     */
+    static boolean radixSort(Sorter sorter, int[] a, int low, int high) {
+        int[] b; int offset = low;
+
+        if (sorter == null || (b = (int[]) sorter.b) == null) {
+            b = (int[]) tryAllocate(a, high - low);
+
+            if (b == null) {
+                return false;
+            }
+        } else {
+            offset = sorter.offset;
+        }
+
+        int start = low - offset;
+        int last = high - offset;
+
+        int[] count1 = new int[256];
+        int[] count2 = new int[256];
+        int[] count3 = new int[256];
+        int[] count4 = new int[256];
+
+        for (int i = low; i < high; ++i) {
+            count1[ a[i]         & 0xFF]--;
+            count2[(a[i] >>>  8) & 0xFF]--;
+            count3[(a[i] >>> 16) & 0xFF]--;
+            count4[(a[i] >>> 24) ^ 0x80]--;
```

Comment:
It seems that C2 can't eliminate the bounds check here because of the `xor`, even though this can't possibly exceed 256. The three masked accesses above are all eliminated. Maybe someone could look in to improving that.

Reply:
https://bugs.openjdk.java.net/browse/JDK-8267332

Reply:
I agree with Laurent (bourgesl), see his comment on May 15 regarding to xor:
using Unsafe is only 2% faster, not worth the extra complexity for few percent.

Reply:
That bounds check shouldn't be there, it's an obvious case and will be fixed: https://github.com/openjdk/jdk/pull/4136

Reply:
Thanks for fixing hotspot C2 (xor) so quickly !